File: src/Physics/aleMeshHydroHLLC.py
# Copyright (C) 2026  Cody Raskin

from PYB11Generator import *
from hydro import *

@PYB11template("dim")
class ALEMeshHydroHLLC(Hydro):
    def pyinit(self,
               nodeList="NodeList*",
               constants="PhysicalConstants&",
               eos="EquationOfState*",
               mesh="Mesh::ALEMesh<%(dim)s>*",
               nodeVelocities=("NodeList*", "nullptr")):
        "nodeVelocities: optional node-centered NodeList (one entry per mesh node) with a Vector 'velocity' field prescribing mesh motion. Omit (or pass None) for a static (Eulerian) mesh."
        return

ALEMeshHydroHLLC2d = PYB11TemplateClass(ALEMeshHydroHLLC,
                              template_parameters = ("2"),
                              cppname = "ALEMeshHydroHLLC<2>",
                              pyname = "ALEMeshHydroHLLC2d",
                              docext = " (2D).")

File: src/Physics/aleMeshHydroHLLC.cc
// Copyright (C) 2026  Cody Raskin

#pragma once
#include "aleMeshHydroBase.hh"
#include "HLL.cc"

// First-order HLLC on ALEMesh: the exact contact-preserving Riemann solve,
// with no slope reconstruction (that needs face-adjacency-chain walking to
// find a "next" cell along a direction, which doesn't exist on an
// unstructured mesh yet -- see aleMeshHydroBase.hh and the ALE hydro plan).
template<int dim>
class ALEMeshHydroHLLC : public ALEMeshHydroBase<dim> {
public:
    using typename ALEMeshHydroBase<dim>::Vector;
    using typename ALEMeshHydroBase<dim>::VectorField;
    using typename ALEMeshHydroBase<dim>::ScalarField;

    ALEMeshHydroHLLC(NodeList* nodeList, PhysicalConstants& constants,
                      EquationOfState* eos, Mesh::ALEMesh<dim>* mesh,
                      NodeList* nodeVelocities = nullptr)
        : ALEMeshHydroBase<dim>(nodeList, constants, eos, mesh, nodeVelocities) {}

    virtual std::string name() const override { return "ALEMeshHydroHLLC"; }

    virtual HLLFlux<dim>
    computeFlux(double rhoL, const Vector& vL, double uL, double pL, double csL,
                double rhoR, const Vector& vR, double uR, double pR, double csR,
                const Vector& normal, double faceVelocity) const override {
        return computeHLLCFluxFromStatesALE<dim>(rhoL, vL, uL, pL, csL,
                                                  rhoR, vR, uR, pR, csR,
                                                  normal, faceVelocity);
    }
};
